### PR TITLE
Learning guide performance improvements

### PR DESCRIPTION
**Pull request by Dantemss**
https://www.pivotaltracker.com/story/show/102283428

- Merged Kevin's CLUE calls by page
- Changed CLUE calls to be by pools
- Improved learning guide performance by combining all calls to Biglearn for each period into 1 call
- Added `all_exercises_pool` to chapters and pages.
- Migration rake task to add `all_exercises_pool` to existing pages and chapters
- Clue caching with key-based expiration, plus time-based expiration.

### CODE REVIEW

**Comment by Dantemss** [comment]
Includes https://github.com/openstax/tutor-server/pull/596

**Comment by Dantemss** [comment]
This is the first part of https://www.pivotaltracker.com/story/show/102283428 (the other part is caching the CLUE's).

**Comment by Dantemss** [comment]
I'll make the historical CLUE's based on this PR, but it can probably be merged independently.
This PR by itself should bring some pretty significant performance improvements (tested manually by causing the fake Biglearn client to sleep for 0.2 seconds on each call).

This PR should be tested with the real Biglearn in dev (or QA).

**Comment by jpslav** [comment]
@kevinburleigh75 can you review?

File: app/external/openstax/biglearn/v1/real_client.rb [hunk 1]
```diff
@@ -90,60 +88,108 @@ def get_projection_exercises(role:, pools: nil, tag_search: nil,
...
-    when Hash
-      raise IllegalArgument, "too many hash conditions" if tag_search.size != 1
-      stringify_tag_search_hash(tag_search.first)
-    when String
-      '"' + tag_search + '"'
-    else
-      raise IllegalArgument
+  def get_clues(roles:, pools:)
+    raise "At least one role must be specified when getting a CLUE" if roles.blank?
+    raise "At least one pool must be specified when getting a CLUE" if pools.blank?
```
Comment (jpslav) [inline app/external/openstax/biglearn/v1/real_client.rb]: If these checks apply to the real BigLearn, they should also apply to the fake client, so I'd suggest moving them up to the `V1` wrapper.

**Comment by jpslav** [comment]
@Dantemss you are the man.  Looks good to me.  I think @kevinburleigh75 is going to test on -kev on Monday.  After that, can merge.

File: app/external/openstax/biglearn/v1/real_client.rb (hunk 1 above)
Comment (Dantemss) [inline app/external/openstax/biglearn/v1/real_client.rb]: Makes sense.

**Comment by Dantemss** [comment]
@jpslav @kevinburleigh75 this is ready for more testing once Biglearn is working. Learning guide should be working properly now.

**Pull request merged**